### PR TITLE
feat(expr): add `array_sum`

### LINKED ISSUE #11996: Implement array_{min/max/sum/sort}

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

add `array_sum` & resolve #11996

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xzhseh** [comment]
The current implementation only support `Int16`, `Int32`, `Int64`.
Though have tried many times, I could only come up with the current workaround, it's dirty and could definitely be refactored.
Also, seems that we should only support `int`, `float`, `decimal` for sum operation?
Any advice is welcomed!

**Comment by wangrunji0408** [comment]
> The current implementation only support `Int16`, `Int32`, `Int64`.
> Though have tried many times, I could only come up with the current workaround, it's dirty and could definitely be refactored.
> Also, seems that we should only support `int`, `float`, `decimal` for sum operation?
> Any advice is welcomed!

I have refactored the function with some type exercise tricks. 🤪

**Comment by xiangjinwu** [comment]
Is this supposed to be consistent with the agg version of `sum`? Note the return type is wider than input type for smallint, int, and bigint. See [PostgreSQL](https://www.postgresql.org/docs/15/functions-aggregate.html#:~:text=sum%20(%20smallint%20)%20%E2%86%92%20bigint) and [RisingWave](https://github.com/risingwavelabs/risingwave/blob/83637cdfbe46c20325d24f6fedc56e3338eea99b/src/expr/src/agg/general.rs#L23-L26).

**Comment by wangrunji0408** [comment]
> Is this supposed to be consistent with the agg version of `sum`? Note the return type is wider than input type for smallint, int, and bigint. See [PostgreSQL](https://www.postgresql.org/docs/15/functions-aggregate.html#:~:text=sum%20(%20smallint%20)%20%E2%86%92%20bigint) and [RisingWave](https://github.com/risingwavelabs/risingwave/blob/83637cdfbe46c20325d24f6fedc56e3338eea99b/src/expr/src/agg/general.rs#L23-L26) (the `sum(int64) -> int64` variant is for 2-phase agg and should not be user visible).

Yes I think it's better to be consistent with agg.

**Comment by xzhseh** [comment]
I don't quite get the point how to be consistent with the `sum` in aggregation 🥺, should we also change to the two-phase building for `array_sum` just like agg?

**Comment by xiangjinwu** [comment]
> I don't quite get the point how to be consistent with the `sum` in aggregation 🥺, should we also change to the two-phase building for `array_sum` just like agg?

I meant the user-visible behavior. We would like `select ..., array_sum(arr), ... from ... where ...` to be equivalent to `select ..., (select sum(elem) from unnest(arr) as foo(elem)), ... from ... where ...`

More specifically, `array_sum(array[32767::smallint, 2::smallint])` shall return `32769` of `bigint` type rather than reporting smallint out of range error.

**Review by wangrunji0408 — COMMENTED** [review]
IIUC, we should change the code in these place in order to align with sum agg:

File: src/expr/src/vector_op/array_sum.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+use risingwave_common::types::{CheckedAdd, ScalarRefImpl};
+use risingwave_expr_macro::function;
+
+use crate::{ExprError, Result};
+
+#[function("array_sum(list) -> *int")]
+#[function("array_sum(list) -> *float")]
+#[function("array_sum(list) -> decimal")]
+#[function("array_sum(list) -> interval")]
+fn array_sum<T>(list: ListRef<'_>) -> Result<T>
```
[the remark this answers is not in the document]
Comment (xiangjinwu) [reply src/expr/src/vector_op/array_sum.rs:25]: We do not need `array_sum(int64[]) -> int64`. Right?
It is actually:
```
array_sum(int16[]) -> int64
array_sum(int32[]) -> int64
```
falling into:
```
array_sum(list) -> int64
```
and
```
array_sum(int64[]) -> decimal
array_sum(decimal[]) -> decimal
```
falling into:
```
array_sum(list) -> decimal
```

**Comment by xzhseh** [comment]
New version updated. Now it should be conform with the agg except `Int256`, seems that the return type of `Int256` in the trait bound can not be satisfied 🤔️?

File: src/expr/src/vector_op/array_sum.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    if flag != 0 {
+        match flag {
+            1 => {
+                let mut sum = 0;
+                for e in list.iter().flatten() {
+                    sum += match e {
+                        ScalarRefImpl::Int16(v) => v as i64,
+                        ScalarRefImpl::Int32(v) => v as i64,
+                        _ => panic!("Expect ScalarRefImpl::Int16 or ScalarRefImpl::Int32"),
+                    }
```
Comment (wangrunji0408) [inline src/expr/src/vector_op/array_sum.rs:54]: Also needs `checked_add` to handle overflow.

File: src/expr/src/vector_op/array_sum.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            }
+            2 => {
+                let mut sum = Decimal::Normalized(0.into());
+                for e in list.iter().flatten() {
+                    sum = sum.add(match e {
+                        ScalarRefImpl::Int64(v) => Decimal::Normalized(v.into()),
+                        ScalarRefImpl::Decimal(v) => v,
+                        // FIXME: We can't panic here due to the macro expansion
+                        _ => Decimal::Normalized(0.into()),
+                    });
```
Comment (wangrunji0408) [inline src/expr/src/vector_op/array_sum.rs:66]: ditto: `checked_add`

**Review by wangrunji0408 — COMMENTED** [review]
Rest LGTM!

**Review by wangrunji0408 — APPROVED** [review]

**Pull request merged**